packages/react/src/chunk.js

```diff
@@ -0,0 +1,70 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import {REACT_CHUNK_TYPE, REACT_MEMO_TYPE} from 'shared/ReactSymbols';
+
+import warningWithoutStack from 'shared/warningWithoutStack';
+
+opaque type Chunk<Props>: React$AbstractComponent<
+  Props,
+  null,
+> = React$AbstractComponent<Props, null>;
+
+export default function chunk<Args, Props, Data>(
+  query: (...args: Args) => Data,
+  render: (props: Props, data: Data) => React$Node,
+): (...args: Args) => Chunk<Props> {
+  if (__DEV__) {
+    if (typeof query !== 'function') {
+      warningWithoutStack(
+        false,
+        'Chunks require a query function but was given %s.',
+        query === null ? 'null' : typeof query,
+      );
+    }
+    if (render != null && render.$$typeof === REACT_MEMO_TYPE) {
+      warningWithoutStack(
+        false,
+        'Chunks require a render function but received a `memo` ' +
+          'component. Use `memo` on an inner component instead.',
+      );
+    } else if (typeof render !== 'function') {
+      warningWithoutStack(
+        false,
+        'Chunks require a render function but was given %s.',
+        render === null ? 'null' : typeof render,
+      );
+    } else {
+      warningWithoutStack(
+        // Do not warn for 0 arguments because it could be due to usage of the 'arguments' object
+        render.length === 0 || render.length === 2,
+        'Chunk render functions accept exactly two parameters: props and data. %s',
+        render.length === 1
+          ? 'Did you forget to use the data parameter?'
+          : 'Any additional parameter will be undefined.',
+      );
+    }
+
+    if (render != null) {
+      warningWithoutStack(
+        render.defaultProps == null && render.propTypes == null,
+        'Chunk render functions do not support propTypes or defaultProps. ' +
+          'Did you accidentally pass a React component?',
+      );
+    }
+  }
+  return function(): Chunk<Props> {
+    let args = arguments;
+    return {
+      $$typeof: REACT_CHUNK_TYPE,
+      query: function() {
+        return query.apply(null, args);
+      },
```

Comment:
In Flight, this function will be replaced by one that suspends on the data, as well as as a render that suspends on the code.